azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -167,6 +167,17 @@ def _run_op(self, op):
 
             op.complete()
 
+        elif isinstance(op, pipeline_ops_base.ShutdownPipelineOperation):
+            # TODO: Make sure the transport is disconnected. This is trickier than it seems.
```

Comment:
># TODO: Make sure the transport is disconnected. This is trickier than it seems. [](start = 12, length = 80)

Aren't you already doing this at a higher level?  There's code in sync_client.py to disconnect before shutting down.

Also, what does "Make sure" mean?  It could mean a few different things.  A few that come to mind:
1. review the transport code to make sure all the shutdown code paths also cause the disconnect.
2. fail this operation if the transport is still connected.
3. cause the transport to disconnect if it's still connected
4. wait for the transport to disconnect if it's still connected. 

For what it's worth, there isn't much state management in this level and below.  e.g. if you try to disconnect, there's no code in here to verify that you're actually connected.  Likewise if you try to publish.  The "make sure the transport is in the correct state" logic happens above this.

Reply:
This is an outdated comment. I have removed it. It referred to the connection in the Transport, but that's been resolved.